### ISSUE TITLE
Performance and qos improvements

### ISSUE THREAD

**Issue by 0anubis0** [fanfan42/VFIO-Passthrough-dual-running-systems-on-laptop#1]
VM to host framebuffer display +mouse/keyboard integration: https://looking-glass.io/ 
Its really fast and good to use. Some note to usage:
-if you are not running your windows with headless nvidia config you need a dummy hdmi
-check the mouse and keyboard config, it has a good config for 5 button mouse
-in windows change the touchpad sensitivity setting to maximum. The low sensitivity mean the touch will be disabled on keypress.

I am using the host with the integrated gpu only, so I disabled the gpu change on startup. With integrated card I dont have HDMI output on host.
To solve this I am planning to create a virtual display on the host and connect to it from the vm with vnc for example. 

Thanks for the repo, it was really useful.
I will post my config later if interested

**Comment by fanfan42** [comment]
Hello, a big Thank You for helping this repository

I will assume you took the Lenovo Legion 5 17ACH6H for your comment. Indeed, I prefer using Switchable Graphics in the BIOS configuration (And I added this mention in the README). I don't always play and mainly work/code on that computer. Using only the integrated GPU in my AMD Ryzen CPU and disconnect the graphics card allows me to save energy (and battery). But yes, even if the Legion series have a good MUX, the CPU is only connected to the integrated screen. With my use case, you can have the second screen by launching the VM and shutting it down just after

I never tried to configure the looking glass, I read many comments on Reddit complaining about the not so good graphics performance on the VM and cpu usage on the Host. But I never gave it a chance. I found a new blog talking about looking-glass installation (https://blog.karaolidis.com/vfio/) and I need to reinstall my VM so I think I will give it a try and test by myself (and add the configuration on this repository). Your config could be a big plus :)

**Comment by 0anubis0** [comment]
Hello, 
Yes, I am using Legion 5 17ACH6H same as you.
I have tried vnc, nomachine, rdp and only the looking glass was capable for gaming.
The lglass has the shared memory interface and also a backup spice interface.
For the install check the homepage. Combined with your config its whole.
It hadnt any performance issue, it was like a real windows. The only thing, it can have coredump sometimes but you can restart it less than a sec. Also this error is special for me probably because I dont have enough ram and using swap.

Another interesting option, I am using zfs file system. It has software raid, compression and really good backup options. The backups and rollback done in a sec so its really good for experimenting. The cons are it eats half of your ram and relatively hard to install.

Just finished the virtual display, I will have my config in this week and upload it :)

**Comment by fanfan42** [comment]
As soon as i will reinstall my VM, I will try some tests on the VM, with and without looking-glass. I think the window tool and some tools from Phoronix will help discovering the real performance on this PC

ZFS is really interesting, I work with this filesystem since 2010, far before it became openzfs. Shadow uses it to provide LUN ISCSI filesystems to their virtual desktop but I think it's a waste of RAM if ZFS is on the same host as the hypervisor. But you give me ideas to use my personal home server, provide zfs LUN and separate the resources for storage and virtualization. For your RAM problems, did you reduce in the BIOS the UMA Frame buffer size to 1G ? By default, it takes 4G when you buy the PC

**Comment by 5S-PETER-PAZMANDI** [comment]
As I see I dont have to sell the ZFS, you know it well :)

This was my first ZFS implementation so I am sure its not the most efficient. I have two 1T SSD partitioned to ZFS raid1 and 50G spare ext4/swap
I have separated the ZFS to root, storage, vm and docker dataset. My main point to use ZFS on root is redundancy and backup. I need the possibility to complete rollback, even for my VM-s. Actually I am using the rollback most of the time for my windows VM.

**One note:
When you reinstalling windows, switching on hyperV support can brick it.
This setting: Enable-WindowsOptionalFeature -Online -FeatureName Microsoft-Hyper-V -All**

About the RAM thanks for reminding me for the UMA buffer, completely forgot it. I can reduce the ZFS memory allocation as well but I will have a ram upgrade in this month.

What did you used for virtual display? Do you have wayland or xorg? I have used evgl driver or xrandr magic for xorg but its really glitchy. These dummy drivers are not so good, next I am trying the AMD or AMDpro driver.

**Comment by fanfan42** [comment]
Your ZFS configuration is a choice. Using it is a loss of performance. In my case, I want to have the best performance possible. In your case, your main point is security of your data, if my disk dies, I lose everything.

I reinstalled a new VM, added some new configuration and added your tip  on Hyper-V activation, it doesn't work anymore for me.

In my case, I don't use any virtual display, my dGPU is directly connected to an extermal display. That's why I think looking-glass won't add any performance upgrade but it's still needed when you don't have an external display. The only virtual display I use is for my lenovo thinkpad and toshiba satellite. It's based on the "old" Intel gvt-g combined with opengl. I still use Xorg. Never configured anything about xrandr or evgl nor AMD or AMDpro driver.

I received today my dummy HDMI plugin. I will try ASAP the configuration with looking glass

**Comment by 0anubis0** [comment]
I am sure the lglass wont be better than external monitor yes.
I have researched the virtual display part, the following what I have found for **linux:**
https://wiki.archlinux.org/title/Extreme_Multihead
**Hardware dependent solutions for integrated intel and amd:**
 -Intel solution works as I heard (3.3.1 in link)
 -I not tested the amd but from what I read it switches the xorg to headless mode what not good for us.
**Hardware independent solutions:**
 -xorg dummy driver (3.3.4) headless mode, not good
 -evdi driver (3.3.3) worked well, thats what I am using now.

**Windows:**
I have found nothing. There is a shitty headless mode for one of the windows pack but not even a clue.
Have to use a **hardware hdmi dummy**, in other name hdmi emulator.
Like this: https://www.amazon.com/Headless-Emulator-Headless-1920x1080-Generation-60Hz%EF%BC%89-3Pack/dp/B06XSY9THQ
Its price around 5 eu, dont buy it for more than 40 eu :)

**Using evdi:**
yay evdi
echo "options evdi initial_device_count=2" > /etc/modprobe.d/local-evdi.conf
reboot

**sethdmi_script:**
xrandr --addmode DVI-I-1-1 1920x1080
xrandr --setprovideroutputsource 1 0 --setprovideroutputsource 2 0
xrandr --output DVI-I-1-1 --mode 1920x1080 --right-of eDP-1
x11vnc --loop -deferupdate 10 -deferptrupdate 10 -display :0 -clip xinerama1

optionally add this option -noxdamage if the display is glitchy, it can be buggy as I heard but works for me
Also looking-glass can have interference with the virtual display, so close it after vnc connection setup. 

**removehdmi_script:**
xrandr --output DVI-I-1-1 --off

I am using **x11vnc server** and **ultravnc client**. Its laggy and not optimal, but usable. I will try other screencapture methods later.
The following client config was good for me:
-hextile encoding
-view only, no input
-fit to screen, no scrollbars
-start in fullscreen mode
-uncheck show toolbar
-optionally you can change the full color to 256 for performance but it will be ugly

**Looking-glass:**
shortcuts: https://looking-glass.io/wiki/Client/Keyboard_shortcuts
Client command what I am using: 
looking-glass-client -m 110 -F input:grabKeyboardOnFocus input:rawMouse input:autoCapture -p 5910
-m 110: escape key put to insert instead of scroll lock

**Comment by fanfan42** [comment]
Hi,

I tested what you did and made a bunch of commits  with looking glass installed on guest and host
I took your command line to launch the looking-glass client, Thanks

I don't understand the other points. Especially about VNC, evdi and set_hdmi script. Is this another way to go on the Guest or something to add to the host to improve performance ?

I let you look what I added. Thanks again

**Comment by 0anubis0** [comment]
The vnc, evdi, and set_hdmi is for a completely different problem. I dont
have a hdmi output with integrated only config. I want to skip the window
manager reload what needed for hybrid mode. I am using the VM with gpu
passthrough to share the hdmi and display port. The trick is I have a
virtual screen on linux and the windows can connect to it via VNC. The
windows VNC client is opened as fullscreen on the windows hdmi screen.
There must be a better solution what I will check later.

I will check your work on the weekend, hope you like the looking glass :)

**Comment by 0anubis0** [comment]
Hi!

I have some experiences since my last answer. I was using KDE plasma and the VM shutdown caused slowdown on the host. I tried to track the error but I didnt find anything except the nvidia driver triggered. I deleted the nvidia driver and changed the desktop to mate. It was more stable and the host slowdown disappeared.
The current problems what I am trying to solve is these:
-If I restart the VM it causes performance issue on it. The only solution to this is the host restart. Did you have this issue?
-After the shutdown, there is an increased power consumption. 
-Did you tried to hibernate or sleep the VM?

**Comment by fanfan42** [comment]
Hello,

Sorry, it's been a long long time since I used the KDE environment. Despite its really nice environment, it takes many resources to run. Mate is not the lightest environment I use/know but it's the most stable I've ever used since 2005 (when it was still gnome 2).

For your other problems, it's funny because I had some "performance" issues last days. But it was not on my Lenovo Legion, only on my lenovo thinkpad L14. In my case, the fan litterally stopped working making my CPU overheating and slowed down the guest and the host. In my case, it might be because of an hardware issue, but maybe you can verify on your Legion if it works well on both your fans.

How can I verify (with tools on my Linux) the increase of power consumption ? I will try it but not before Monday next week, I don't have my Legion with me. I think, I have the case but not sure. In fact, my electricity bill warned me last June so maybe I should verify this.

Never tried hibernate or sleep the Guest, I never use hibernation because I don't want to do intense writing on my SSD. And don't see the interest of sleep for a guest. In fact, if you check my configuration, Hibernate and Sleep are deactivated on each guest

Hope it helps,

Regards

**Comment by 0anubis0** [comment]
For me with the legion 5 I didnt have any fan issue. At least I didnt notice, I will check it.

The power consumption was measured by my battery when I unplugged the power cable. Also noticed it by the increased heating and lower battery life. These acpi modules could help: https://wiki.archlinux.org/title/laptop

As I get from your writing you didnt had this performance issue after restarting the VM. Maybe I should check it without ZFS. I got the feeling its some kind of IO problem but I wasnt able to measure it on windows.

**Comment by fanfan42** [comment]
Hello,

So, I made a test last monday. I started at 7:25 PM when I was on my desktop. I configured the laptop on "on-demand" state (White LED on the power button) and unplugged the energy wire. No WiFI, just Ethernet. Nvidia card was not loaded at start (I never load it normally, only when gaming). My test was to watch the energy state in my battery. The first measurement was during checking news and stuff on Firefox for 5 minutes:

`fanfan@bahamut:~$ cat /sys/class/power_supply/BAT0/energy_now ; cat /sys/class/power_supply/BAT0/power_now ; sleep 300 ; cat /sys/class/power_supply/BAT0/energy_now ; cat /sys/class/power_supply/BAT0/power_now`
78160000 # uWh
11350000 # uW
76480000
13636000

At 7:30 PM, I was playing The Witcher 3 for around 15 minutes. I have to say I never tested playing on battery and I noticed something wrong during my play : poor performance, lag, etc. I didn't change any graphics settings from those I use when plugged on electricity. 

So, a 7:44 PM, here are the battery states for 5 minutes :
`fanfan@bahamut:~$ cat /sys/class/power_supply/BAT0/energy_now ; cat /sys/class/power_supply/BAT0/power_now ; sleep 300 ; cat /sys/class/power_supply/BAT0/energy_now ; cat /sys/class/power_supply/BAT0/power_now`
41250000
28367000
37950000
25150000

During that test and the rest of the battery life, I was still doing stuff on my browser and listening to music using bluetooth speakers. At 8:35 PM, my system told me I had 20% battery life remaining. At 9:08 PM, it was on critical state. At 9:12 PM, last measurement:

`fanfan@bahamut:~$ cat /sys/class/power_supply/BAT0/energy_now ; cat /sys/class/power_supply/BAT0/power_now` 
2490000
24004000

The laptop finally died at 9:15 PM. So. during that test, my battery lasted less than 2 hours which is not really cool but it's a gaming laptop after all and not really a surprise when seeing some testers saying the laptop lasts around 80 minutes when gaming at high performance  (https://www.ultrabookreview.com/50999-lenovo-legion-5-17-review/).

And to be more precise, at the first measurement, I had 78.16 Wh at energy_now and 11.35 W at power_now which gives these remaining times during the tests:
7:25PM: 78.16 Wh / 11.35 W =~ 6,88 hours remaining
7:30PM:  76,48 / 13.636 =~ 5.60 hours remaining
So, even in the first 5 min, without graphic card, I had 15.36 minutes loss on battery life ( (6.88 * 60 - 5.6 * 60) / 5 minutes )

7:44 PM:  41.25 / 28.367 =~ 1.45 hours remaining
7:49 PM:  37.95 / 25.15 =~ 1.51 hours remaining or around 1h30
After a 15 minutes play at high performance (but poor quality graphics :/ ), I already lost ~5 hours on battery life, that's enormous

9:12 PM:  2.49 / 24.004 =~ 0.10 hours remaining or 6 minutes

These tests show me that even if my 15 min play drained all my battery, I could still see a correct battery life for the remaining time. And even with Nvidia loaded which added around 13W consumption.

Hope it will help you
